### PR TITLE
CRIMAPP-1357 caseworker can add MAAT decisions

### PR DESCRIPTION
**Pull request by timpeat**
## Description of change

Caseworker can add a MAAT decision to an application

## Link to relevant ticket
Add decision by USN: [CRIMAPP-1357](https://dsdmoj.atlassian.net/browse/CRIMAPP-1357)
Update a decision based on MAAT changes: [CRIMAPP-1397](https://dsdmoj.atlassian.net/browse/CRIMAPP-1397)
Add a decision by MAAT ID [CRIMAPP-1359](https://dsdmoj.atlassian.net/browse/CRIMAPP-1359)

## Notes for reviewer

We use the MAAT ID for the decision_id of MAAT decisions. This is used to create a unique event stream for the decision and ensure idempotency. (NB: we may start receiving notifications from MAAT relating to decisions.)

## How to manually test the feature

[CRIMAPP-1357]: https://dsdmoj.atlassian.net/browse/CRIMAPP-1357?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CRIMAPP-1397]: https://dsdmoj.atlassian.net/browse/CRIMAPP-1397?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-1359]: https://dsdmoj.atlassian.net/browse/CRIMAPP-1359?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .env.test [hunk 1]
```diff
@@ -15,6 +15,7 @@ MAAT_API_OAUTH_URL='https://oauth.example.com'
 MAAT_API_CLIENT_ID='TestMaatApiClientID'
 MAAT_API_CLIENT_SECRET='TestMaatApiClientSecret'
 MAAT_API_API_URL='https://example.com'
+MAAT_API_STARTING_MAAT_ID=3
```
Comment (timpeat) [inline .env.test:18]: This constant is set on a per environment basis and is used to prevent unsupported applications being retrieved.

File: app/aggregates/deciding/command.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module Deciding
   class Command < Dry::Struct
-    attribute :decision_id, Types::Uuid
+    attribute :decision_id, Types::Uuid | Types::Integer
```
Comment (timpeat) [inline app/aggregates/deciding/command.rb:3]: We use the MAAT ID for the decision_id of MAAT decisions.

File: app/aggregates/deciding/decision.rb [hunk 3]
```diff
@@ -6,37 +6,67 @@ def initialize(decision_id)
       @decision_id = decision_id
     end
 
-    attr_reader :application_id, :decision_id, :funding_decision, :comment, :state
+    attr_reader :application_id, :decision_id, :funding_decision, :comment,
+                :state, :reference, :maat_id, :checksum, :case_id
+
+    # For decisions entered on CrimeReview by the caseworker
```
Comment (timpeat) [inline app/aggregates/deciding/decision.rb:12]: TYPO

File: app/aggregates/deciding.rb [hunk 4]
```diff
@@ -1,11 +1,28 @@
...
+        new(data: args.merge(data))
+      end
+    end
+  end
 
   class DraftCreated < RailsEventStore::Event; end
-  class InterestsOfJusticeSet < RailsEventStore::Event; end
-  class FundingDecisionSet < RailsEventStore::Event; end
-  class CommentSet < RailsEventStore::Event; end
+  class DraftCreatedFromMaat < RailsEventStore::Event; end
```
Comment (timpeat) [inline app/aggregates/deciding.rb:21]: All of these should use Event, just don't call with build the creating events.

File: app/aggregates/deciding/commands/create_draft.rb [hunk 5]
```diff
@@ -2,10 +2,11 @@ module Deciding
   class CreateDraft < Command
     attribute :application_id, Types::Uuid
     attribute :user_id, Types::Uuid
+    attribute :reference, Types::Integer
 
     def call
       with_decision do |decision|
-        decision.create_draft(user_id:, application_id:)
+        decision.create_draft(user_id:, application_id:, reference:)
```
Comment (timpeat) [inline app/aggregates/deciding/commands/create_draft.rb:9]: Including reference in all deciding events simplifies linking to streams based on reference for submission history.

File: app/aggregates/deciding/decision.rb [hunk 6]
```diff
@@ -49,8 +79,26 @@ def set_comment(user_id:, comment:)
...
+
+    def means
+      return if @means.nil?
+
+      Types::MeansDecision[@means]
+    end
+
     def interests_of_justice
-      return {} if @interests_of_justice.nil?
+      return if @interests_of_justice.nil?
```
Comment (timpeat) [inline app/aggregates/deciding/decision.rb:101]: Review this

File: app/aggregates/reviewing/review.rb [hunk 7]
```diff
@@ -5,6 +5,7 @@ class Review
     def initialize(id)
       @id = id
       @decision_ids = []
+      @maat_ids = []
```
Comment (timpeat) [inline app/aggregates/reviewing/review.rb:8]: remove maat_ids for now

File: app/aggregates/reviewing/review.rb [hunk 8]
```diff
@@ -52,10 +53,16 @@ def mark_as_ready(user_id:)
       apply MarkedAsReady.build(self, user_id:)
     end
 
-    def add_decision(user_id:, decision_id:)
+    def add_decision(decision_id:, user_id: nil)
       apply DecisionAdded.build(self, user_id:, decision_id:)
     end
 
+    def add_maat_decision(decision_id:)
```
Comment (timpeat) [inline app/aggregates/reviewing/review.rb:60]: Are using this?

File: app/components/decision_component.rb [hunk 9]
```diff
@@ -3,71 +3,60 @@ class DecisionComponent < ViewComponent::Base
   include AppTextHelper
   include ComponentsHelper
 
-  def initialize(decision:, decision_iteration: nil, editable: false)
+  def initialize(decision:, decision_iteration: nil, show_actions: false)
```
Comment (timpeat) [inline app/components/decision_component.rb:6]: confirm show_actions consistent

File: app/controllers/casework/decisions_controller.rb [hunk 10]
```diff
@@ -10,7 +10,8 @@ def create
       args = {
         application_id: @crime_application.id,
         user_id: current_user_id,
-        decision_id: decision_id
+        decision_id: decision_id,
+        reference: @crime_application.reference
```
Comment (timpeat) [inline app/controllers/casework/decisions_controller.rb:14]: We're recording reference to events to aid linking to submission history

File: app/controllers/casework/maat_decisions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  class MaatDecisionsController < Casework::BaseController
+    # Used for managing linked decisions and manually linking a decision by MAAT ID.
+    # See DefaultMaatDecisionsController
+    include SetDecisionAndAuthorise
+
+    before_action :set_decision, :require_maat_decision!, except: [:create, :new, :create_by_reference]
+
+    def new
+      @form_object = ::Decisions::MaatIdForm.new(
+        application_id: @crime_application.id
```
Comment (timpeat) [inline app/controllers/casework/maat_decisions_controller.rb:11]: Does this need reference?

File: app/lib/types.rb [hunk 12]
```diff
@@ -64,23 +64,24 @@ module Types # rubocop:disable Metrics/ModuleLength
...
+  TestResult = String.enum('pass', 'fail', 'inel', 'full')
+  MeansResult = TestResult
+  InterestsOfJusticeResult = String.enum(TestResult['pass'], TestResult['fail'])
+
+  FundingDecisionResult = String.enum(*%w[granted granted_on_ioj fail_on_ioj failmeans failioj])
 
   InterestsOfJusticeDecision = Hash.schema(
     result: InterestsOfJusticeResult,
-    details?: String,
+    details?: String.optional,
```
Comment (timpeat) [inline app/lib/types.rb:75]: Update in Schema Gem?

File: app/models/concerns/decision_form_persistance.rb [hunk 13]
```diff
@@ -24,14 +31,24 @@ def permit_params(params = {})
...
   private
 
+  def raise_if_read_only!(attributes)
+    attributes.each_key do |attr_name|
+      next unless self.class.readonly_attribute?(attr_name.to_s)
+
+      raise ActiveRecord::ReadonlyAttributeError, attr_name
+    end
+  end
+
```
Comment (timpeat) [inline app/models/concerns/decision_form_persistance.rb:51]: ☝️ prevent setting of anything other than editable prams being updated.

File: app/lib/maat/decision.rb [hunk 14]
```diff
@@ -2,11 +2,18 @@ module Maat
...
-    attribute :funding_decision, Types::FundingDecisionResult
+    attribute? :case_id, Types::String.optional
+    attribute :interests_of_justice, Types::InterestsOfJusticeDecision.optional
+    attribute :means, Types::MeansDecision.optional
+    attribute :funding_decision, Types::FundingDecisionResult.optional
     attribute? :comment, Types::String.optional
 
+    def checksum
+      Digest::MD5.hexdigest(to_json)
+    end
```
Comment (hiboabd) [inline app/lib/maat/decision.rb:13]: Maybe a silly question but why is this required?

Comment (timpeat) [reply app/lib/maat/decision.rb:13]: We store this in Review and use it to find out if the decision on MAAT has changed since we last updated the Review record here:

https://github.com/ministryofjustice/laa-review-criminal-legal-aid/pull/726/files#diff-2e01c001095fbec6bcf1621ac7e2de1b89af757b5f2e2b88549688e6ba5bf725R32

File: .env.test (hunk 1 above)
Comment (timpeat) [reply .env.test:18]: Rename constant and add explanation

File: config/kubernetes/staging/config_map.yml [hunk 15]
```diff
@@ -14,3 +14,7 @@ data:
   OMNIAUTH_AZURE_CLIENT_ID: 5d1ba28b-234a-44e1-9464-f5bec5228bfc
   OMNIAUTH_AZURE_TENANT_ID: c6874728-71e6-41fe-a9e1-2e8c36776ad8
   OMNIAUTH_AZURE_REDIRECT_URI: https://staging.review-criminal-legal-aid.service.justice.gov.uk/users/auth/azure_ad/callback
+  MAAT_API_API_URL: https://laa-crime-applications-adaptor-uat.apps.live.cloud-platform.service.justice.gov.uk
+  MAAT_API_OAUTH_URL: https://caa-api-uat.auth.eu-west-2.amazoncognito.com
+  MAAT_API_STARTING_MAAT_ID: 6039300
```
Comment (timpeat) [inline config/kubernetes/staging/config_map.yml:19]: Rename: 'earliest supported maat_id' ?

File: app/aggregates/deciding/decision.rb [hunk 16]
```diff
@@ -6,37 +6,69 @@ def initialize(decision_id)
       @decision_id = decision_id
```
Comment (timpeat) [inline app/aggregates/deciding/decision.rb:6]: Add comment explain the decision_id.

File: app/controllers/casework/maat_decisions_controller.rb [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+      )
+    end
+
+    # Used to create a decision when a caseworker needs to link to a decision
+    # not imported directly from CrimeApply (e.g., for split cases or when an
+    # import failed due to a technical issue).
+    def create
```
Comment (timpeat) [inline app/controllers/casework/maat_decisions_controller.rb:18]: Handle 404 from MAAT

File: spec/system/casework/deciding/adding_a_maat_decision_by_maat_id_spec.rb [hunk 18]
```diff
@@ -0,0 +1,128 @@
+require 'rails_helper'
+
+RSpec.describe 'Adding a decision by MAAT reference' do
```
Comment (timpeat) [inline spec/system/casework/deciding/adding_a_maat_decision_by_maat_id_spec.rb:3]: C/P error. Update to MAAT ID

File: app/lib/maat/get_decision.rb [hunk 19]
```diff
@@ -26,8 +35,8 @@ def get(path)
 
       # NOTE: The MAAT API returns a 200 status even if a decision has not been found.
       # The existence of the decision can only be determined by checking if
-      # the response body is empty.
-      return nil if response.body.empty?
+      # the response body includes a MAAT ID (maat_ref).
+      return nil unless response.body.present? && response.body['maat_ref'].present?
```
Comment (timpeat) [inline app/lib/maat/get_decision.rb:39]: Handle 404/403

File: config/settings.yml [hunk 20]
```diff
@@ -24,7 +24,7 @@ feature_flags:
     staging: true
     production: true
   adding_decisions:
-    local: false
+    local: true
```
Comment (timpeat) [inline config/settings.yml:27]: Change to false
